python/tvm/relay/op/strategy/x86.py

```diff
@@ -627,16 +627,16 @@ def batch_matmul_strategy_cpu(attrs, inputs, out_type, target):
     if (
         not attrs.transpose_a
         and attrs.transpose_b
-        and target_has_vnni(mcpu)
+        and target_has_avx512(mcpu)
         and inputs[0].dtype == "uint8"
         and inputs[1].dtype == "int8"
         and inputs[1].shape[-2] % 16 == 0
         and inputs[1].shape[-1] % 4 == 0
     ):
         strategy.add_implementation(
-            wrap_compute_batch_matmul(topi.x86.batch_matmul_vnni_compute, need_out_dtype=True),
-            wrap_topi_schedule(topi.x86.schedule_batch_matmul_vnni),
-            name="batch_matmul_vnni.x86",
```

Comment:
* Don't remove the ```vnni``` one completley
* Instead, maybe a descend strategy would be better:
         -  if ```has_vnni```
         - elsif ```has_avx512```
         - elsif ```has_avx2``` (in a next PR?)
         - elsif ```has_ssse3``` (in a next PR?)

Reply:
Hello @cbalint13! Thank you for your nits and remarks! In this case VNNI was not removed but extended, as you know VNNI is a part of AVX512 architectures. The fork is here:
https://github.com/apache/tvm/blob/main/python/tvm/topi/x86/tensor_intrin.py#:~:text=def%20dot_16x1x16_uint8_int8_int32()%3A,return%20dot_16x1x16_uint8_int8_int32_skylake()
As you correctly remarked avx2 and ssse3 are also processed here, but they are not accessable due to high-level check target_has_avx512. Possibly you suggestion is good way how to resolve it further. Now I only extended existed approach for avx512.

Reply:
>  know VNNI is a part of AVX512 architectures. The fork is here:
>https://github.com/apache/tvm/blob/main/python/tvm/topi/x86/tensor_intrin.py#:~:text=def%20dot_16x1x16_uint8_int8_int32()%3A,return%20dot_16x1x16_uint8_int8_int32_skylake()
> As you correctly remarked avx2 and ssse3 are also processed here, but they are not accessable due to high-level check target_has_avx512.

Can't see original **llvm.x86.vnni** instrinsic one in the above.
This switch, right in the provided **tensor_intrin.py** fork:
```
    if target_has_vnni(mcpu):
        # VNNI capable platform
        return dot_16x1x16_uint8_int8_int32_cascadelake()
    # vpmaddubsw/vpmaddwd fallback
    return dot_16x1x16_uint8_int8_int32_skylake()
```

As +SIMD keeps coming, would't be better to stay as upcoming [strategy/x86.py](https://github.com/apache/tvm/blob/dd1eb2498a4e2c867a16d7d9ccea010c396e10ae/python/tvm/relay/op/strategy/x86.py#L594-L621)  ```if/elif``` + preferential ```plevel``` ?
 * The **tensor_intrin.py** would remain only a enumeration of SIMD (no decisions), triages would stay **strategy**, etc.
 * User may control the fall into right strategy by narrowing ```"llvm +mattr={avx512bw,avxvnni,...}"```, as [llvm flags](https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/BuiltinsX86.def).

Reply:
Hello @cbalint13! Your view looks reasonable and there is no problems to reimplement it from my side. But I did not implement method `dot_16x1x16_uint8_int8_int32` with conditions on tensor_intrin.py side and thought that it is brick to build some concept. @elvin-n and @jwfromm what do you think about Balint's view?

Reply:
> Hello @cbalint13! Your view looks reasonable and there is no problems to reimplement it from my side. But I did not implement method `dot_16x1x16_uint8_int8_int32` with conditions on tensor_intrin.py side and thought that it is brick to build some concept. @elvin-n and @jwfromm what do you think about Balint's view?

@vvchernov ,

Thanks for clarifications, I see your point, it is perfectly fine way too.
I think by making all CI tests to pass in green more reviewers will come.

---

I try sum up, on this very pinned PR change on **strategy/x86.py**, visibile on top of this thread:
```
-from tvm.topi.x86.utils import target_has_vnni
+from tvm.topi.x86.utils import target_has_avx512

-        and target_has_vnni(mcpu)
+        and target_has_avx512(mcpu)

-        wrap_compute_dense(topi.x86.dense_vnni),
-        wrap_topi_schedule(topi.x86.schedule_dense_vnni),
+        wrap_compute_dense(topi.x86.dense_int8),
+        wrap_topi_schedule(topi.x86.schedule_dense_int8),
```
* This merge vnni to avx512 (under new **dense_int8** umbrella) arguing that VNNI is subset of AVX512 group.
* VNNI is subset of AVX512 group, however there are CPU having AVX512 but **no VNNI** [1].

[1] https://en.wikipedia.org/wiki/AVX-512#VNNI

My view was to leave separate avx512 & vnni(as was) in strategy/x86.py (not to merge vnni->avx512)
My argument was to triage any SIMD right in **strategy/x86.py** as upcoming AMX do [here](https://github.com/apache/tvm/blob/dd1eb2498a4e2c867a16d7d9ccea010c396e10ae/python/tvm/relay/op/strategy/x86.py#L594-L621) + plevel control.
I saw VNNI and AVX512 as potentialy independend things, moreover choosable via "llvm +mattr=...".

Reply:
The instruction set for SSE/AVX2/AVX512 for int8 is absolutely the same, the only difference is the number of lanes. Additionally, the patterns how these int8 instructions (VPMADDUBSW/VPMADDWD/VPADDD) are used, is the same as the only VNNI instruction (VPDPBUSD). I.e. it is reasonable to have the only tvm intrinsic, it is reasonable to remove VNNI from the name of the function, and it is reasonable to extend these intrinsic function to SSE and AVX2 that is not done yet in this PR

Reply:
> The instruction set for SSE/AVX2/AVX512 for int8 is absolutely the same, the only difference is the number of lanes.

   Yes, same class doing integer dot products on immediate registers, but mention:

   - *different* clocking, timing & implementation on ASIC
   - (auto)tensorization opportunities differ as inner loops match *differently*

> Additionally, the patterns how these int8 instructions (VPMADDUBSW/VPMADDWD/VPADDD) are used, is the >same as the only VNNI instruction (VPDPBUSD).

  Right.

   - VNNI insn. accumulates into int32 lanes in single step: vpdpbusd
   - AVX512 (incl. AVX2, SSSE3 ones)  does same in two-step, e.g: pmaddubs + pmadd

> I.e. it is reasonable to have the only tvm intrinsic, it is reasonable to remove VNNI from the name of the function, and it is reasonable to extend these intrinsic function to SSE and AVX2 that is not done yet in this PR

   - Indeed the proposed intrinsic merger is perfectly fine.
   - It was possible to question it with reasonable arguments.

Reply:
> different clocking, timing & implementation on ASIC

What kind of ASIC do you mean?

> (auto)tensorization opportunities differ as inner loops match differently

Under `tensorization opportunities differ` do yo mean different number of lanes for different instruction set which can be reflected in potential different blocking size? Or something else?

Reply:
> > different clocking, timing & implementation on ASIC
> 
> What kind of ASIC do you mean?

* CPU, family of x86, different generations, varying extended ISA layouts: amx avx512 vnni avx2 ssse3 sse2

> > (auto)tensorization opportunities differ as inner loops match differently
> Under `tensorization opportunities differ` do yo mean different number of lanes for different instruction set which can be reflected in potential different blocking size?

* Yes, both input-widths and output-lanes yields different outcomes, varying performances.
* E.g. autotensorizer will opportunistically search to permute & match inner loops to these varying sizes.

> Or something else?

* TVM is a compiler after all, to my knowledge the only one capable of auto-tensorization with arbitrary intrinsic.

Reply:
I would consider amx vs vnni avx512 avx2 sse3 (btw, there is no sse2 for int8, required instructions appeared if I am not mistaken in sse3.x) because first is matrix multiplication, other ones are vector instructions. For now I propose to go from local to generic and when we see needs in differentiate vector sets, we will do this. For now pattern look similar for all of vector instructions, the aspect of blocking should be added separately if it is not done yet, The aspect of lanes in TVM intrinsic should be covered in this PR

> match inner loops to these varying sizes.
The inner loop is the same for all these instructions. It will be 
```
                for (int k = 0; k < 4; k++){
                    output[i] += data[k] * kernel[i][k]
                }
```

> TVM is a compiler after all, to my knowledge the only capable of auto-tensorization with arbitrary intrinsic.

again, I propose to move from local to generic patterns. We do not limit anything for now